src/ui/public/typeahead/typeahead.js

```diff
@@ -1,225 +1,94 @@
-import _ from 'lodash';
-import 'ui/typeahead/typeahead.less';
-import 'ui/typeahead/_input';
-import 'ui/typeahead/_items';
+import './typeahead.less';
+import './typeahead_input';
+import './typeahead_item';
+import template from './typeahead.html';
 import { uiModules } from 'ui/modules';
 import { comboBoxKeyCodes } from '@elastic/eui';
 
+const { UP, DOWN, ENTER, TAB, ESCAPE } = comboBoxKeyCodes;
 const typeahead = uiModules.get('kibana/typeahead');
 
 typeahead.directive('kbnTypeahead', function () {
   return {
-    restrict: 'A',
+    template,
+    transclude: true,
+    restrict: 'E',
     scope: {
-      historyKey: '@kbnTypeahead',
+      items: '=',
+      itemTemplate: '=',
       onSelect: '&'
     },
+    bindToController: true,
     controllerAs: 'typeahead',
+    controller: function () {
+      this.isHidden = true;
+      this.selectedIndex = null;
 
-    controller: function ($scope, PersistedLog, config) {
-      const self = this;
-      self.query = '';
-      self.hidden = true;
-      self.focused = false;
-      self.mousedOver = false;
-
-      self.setInputModel = function (model) {
-        $scope.inputModel = model;
-
-        // watch for changes to the query parameter, delegate to typeaheadCtrl
-        $scope.$watch('inputModel.$viewValue', self.filterItemsByQuery);
+      this.submit = () => {
+        const item = this.items[this.selectedIndex];
+        this.onSelect({ item });
+        this.selectedIndex = null;
       };
 
-      self.setHidden = function (hidden) {
-        self.hidden = !!(hidden);
+      /**
+       * Sets the selected index to the given value. If the value is less than
+       * zero, it will wrap around to the end, and if the value is greater than
+       * the number of items, it will wrap around to the beginning.
+       */
+      this.setSelectedIndex = (index) => {
+        this.selectedIndex = (index + this.items.length) % this.items.length;
       };
```

Comment:
Thanks, I didn't realize `index` could be <0, so this comment makes things more clear. If `index` can be greater than the number of items, can it also be a negative number whose absolute value is greater than the number of items? For example, `(-8 + 6) % 6 = -2`

Reply:
@Bargs I've updated the code so that it reads better. Could you take a look?

Reply:
That makes a lot more sense 😄